### PR TITLE
Add Port and ActivePort support

### PR DESCRIPTION
**Pull request by unai-ndz**
Closes: #16

### CODE REVIEW

File: tests/pulseaudio_dbus_spec.lua [hunk 1]
```diff
@@ -201,4 +207,35 @@ describe("PulseAudio with DBus", function ()
...
+                          assert.is_equal(_sink.ActivePort, ports_array[p].object_path)
+                          return
+                        end
+                      end
+                    end
+                  end
+                  print("\nNOTE: Won't set the next port as the ActivePort because no sink in this machine has more than one port")
+            end)
+
+           it("Will create a monitored proxy", function ()
```
Comment (stefano-m) [inline tests/pulseaudio_dbus_spec.lua:237]: Maybe this test should not be included in this PR?

**Review by stefano-m — CHANGES_REQUESTED** [review]

**Comment by stefano-m** [comment]
Thanks for the PRs @unai-ndz 

I am slowly going through them.

Should they be applied in a specific order? If so, would you mind letting me know which goes first, second, etc. please?

**Comment by unai-ndz** [comment]
They do not depend on each other, but they were developed in the same order as they were posted here. So It may be simpler to merge if done in that order; #13, #15, #17, #18.

File: src/pulseaudio_dbus/init.lua [hunk 2]
```diff
@@ -349,6 +349,24 @@ function pulse.Device:toggle_muted()
...
+-- @see pulse.Device:set_active_port
+function pulse.Device:get_active_port()
+  return self:Get("org.PulseAudio.Device", "ActivePort")
+end
+
+--- Set the active port object path
+-- @tparam string value port object path
+-- @see pulse.Device:get_active_port
+function pulse.Device:set_active_port(value)
+  self:Set("org.PulseAudio.Core1.Device",
```
Comment (stefano-m) [inline src/pulseaudio_dbus/init.lua:364]: Since `value` must be a valid port, can you please add a check using the `Ports` property to ensure that `value` is correct please?

Something like

``` lua
local available_ports = self:Get("org.PulseAudio.Core1.Device", "Ports")
-- assert that value is in available_ports
self:Set(...)
```

File: tests/pulseaudio_dbus_spec.lua [hunk 3]
```diff
@@ -201,4 +207,30 @@ describe("PulseAudio with DBus", function ()
                     end
                   end
            end)
+
+           it("Will set the next port as the ActivePort", function()
```
Comment (stefano-m) [inline tests/pulseaudio_dbus_spec.lua:211]: I think that we should test using both `get_active_port` and `set_active_port` here. The two functions have no coverage at the moment.

File: tests/pulseaudio_dbus_spec.lua [hunk 4]
```diff
@@ -201,4 +207,30 @@ describe("PulseAudio with DBus", function ()
...
+                      for p=1,total_number_of_ports do
+                        if _sink.ActivePort ~= ports_array[p].object_path then
+                          _sink:set_active_port(ports_array[p].object_path)
+                          assert.is_equal(_sink.ActivePort, ports_array[p].object_path)
+                          return
+                        end
+                      end
+                    end
+                  end
+                  print("\nNOTE: Won't set the next port as the ActivePort because no sink in this machine has more than one port")
```
Comment (stefano-m) [inline tests/pulseaudio_dbus_spec.lua:234]: Can you modify the comment saying that the test is being skipped please?

**Review by stefano-m — CHANGES_REQUESTED** [review]
Thanks for the PR, I spotted a bug and suggested a couple of changes.

Looking forward to hearing from you!

**Comment by stefano-m** [comment]
FTR I added travis-ci support (see #19) so we can have some automated tests running. This should help with your tests too once you rebase.

Thank you!

File: src/pulseaudio_dbus/init.lua [hunk 5]
```diff
@@ -349,6 +349,24 @@ function pulse.Device:toggle_muted()
   return self:is_muted()
 end
 
+--- Get the current active port object path
+-- @return the active port object path
+-- @return nil if no active port is set
+-- @see pulse.Device:set_active_port
+function pulse.Device:get_active_port()
+  return self:Get("org.PulseAudio.Device", "ActivePort")
```
[the remark this answers is not in the document]
Comment (unai-ndz) [reply src/pulseaudio_dbus/init.lua:357]: I'm not sure about accepting/returning objects. Getters and properties returning the object path as a string was a little odd when learning the API but ends up being pretty consistent, not only with the rest of pulseadio_dbus but with the pulseaudio docs, etc.
I'm not against it though. Maybe it's a good idea to avoid boilerplate code. But I think it should be done in a future PR, addressing it for all getters/setters in which makes sense. i.g. get/set_fallback_source/sink.

[the remark this answers is not in the document]
Comment (stefano-m) [reply src/pulseaudio_dbus/init.lua]: Yeah, I agree. It might be a nice feature for the future.

File: src/pulseaudio_dbus/init.lua [hunk 6]
```diff
@@ -349,6 +349,39 @@ function pulse.Device:toggle_muted()
...
+  end
+
+  if port_is_valid then
+    self:Set("org.PulseAudio.Core1.Device",
+             "ActivePort",
+             lgi.GLib.Variant("o", value))
+    self.ActivePort = {signature="o", value=value}
+  else
+    -- value it's not a valid port in this device
+    return false
```
Comment (stefano-m) [inline src/pulseaudio_dbus/init.lua:381]: I think that we should throw an error rather than returning a boolean. We could simply do something like

``` lua
assert(port_is_valid, string.format("Invalid port (%s) for device %s", value, self.Name))
```

(and update the docstring) on line 374 rather than `if..else`

We should also add a test please.

**Review by stefano-m — CHANGES_REQUESTED** [review]

File: src/pulseaudio_dbus/init.lua (hunk 6 above)
Comment (unai-ndz) [reply src/pulseaudio_dbus/init.lua:381]: No problem, this actually makes a lot of sense. I don't know what exactly you wanna test for though. If an invalid port fails?

Comment (stefano-m) [reply src/pulseaudio_dbus/init.lua:381]: >I don't know what exactly you wanna test for though.

I think that there should be a test for both a valid and an invalid port. I think we already cover the case for a valid port, so we can simply add a test that tries to set a bogus port and detects the error.

The [Busted Test Framework](http://olivinelabs.com/busted/) has `assert.has_error` which should be helpful.

Comment (unai-ndz) [reply src/pulseaudio_dbus/init.lua]: Thanks for the tip, the test is pretty minimal but I think it should be enough.

**Review by stefano-m — APPROVED** [review]

**Comment by stefano-m** [comment]
Closed by 99a01aa4a8ae054b5ad42c61cbf1553e8ba22280

Original commits have been preserved.

Thanks!
